blockchain/types/derivesha/mux.go

```diff
@@ -33,48 +33,66 @@ type IDeriveSha interface {
 	DeriveSha(list types.DerivableList) common.Hash
 }
 
+type GovernanceEngine interface {
+	ParamsAt(num uint64) (*params.GovParamSet, error)
+}
+
 var (
 	config     *params.ChainConfig
-	instances  map[int]IDeriveSha
+	gov        GovernanceEngine
+	impls      map[int]IDeriveSha
 	emptyRoots map[int]common.Hash
 
 	logger = log.NewModuleLogger(log.Blockchain)
 )
 
 func init() {
-	instances = map[int]IDeriveSha{
+	impls = map[int]IDeriveSha{
 		types.ImplDeriveShaOriginal: DeriveShaOrig{},
 		types.ImplDeriveShaSimple:   DeriveShaSimple{},
 		types.ImplDeriveShaConcat:   DeriveShaConcat{},
 	}
 
 	emptyRoots = make(map[int]common.Hash)
-	for implType, instance := range instances {
-		emptyRoots[implType] = instance.DeriveSha(types.Transactions{})
+	for implType, impl := range impls {
+		emptyRoots[implType] = impl.DeriveSha(types.Transactions{})
```

Comment:
Probably not related this PR, but is this used for emptyReceiptHash as well?

Reply:
I can't find `emptyReceiptHash` in klaytn codebase.
Did you mean
- `types.Header.EmptyReceipt()`? yes it uses EmptyRootHash(num) https://github.com/klaytn/klaytn/blob/dev/blockchain/types/block.go#L150
- `emptyRoot` constants 56e8.. ? no, I assumed it's for state tries, so I didn't make it depend on DeriveShaImpl.

Reply:
@blukat29 Oh, I meant if `impl.DeriveSha(types.Transactions{})` is the same as `impl.DeriveSha(types.Receipts{})` for all derive sha implementations.

Reply:
I did a quick test and it passed.

```go
        for _, impl := range impls {
                assert.Equal(t,
                        impl.DeriveSha(types.Transactions{}),
                        impl.DeriveSha(types.Receipts{}),
                )
        }
```

Reply:
I suppose they are just emptytrie.hash() or keccak256()?

Reply:
Yes, right. It should be okay, but we needed to check the all deriveSha implementations.
It would be better to use more general param instead of `types.Transactions{}` though.
Thanks for updating.